SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestRPCTimeout/SpatialTestRPCTimeout.cpp

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SpatialTestRPCTimeout.h"
+#include "Components/BoxComponent.h"
+#include "GameFramework/Character.h"
+#include "GameFramework/PlayerController.h"
+#include "SpatialFunctionalTestFlowController.h"
+#include "SpatialTestRPCTimeoutGameMode.h"
+#include "SpatialTestRPCTimeoutPlayerController.h"
+
+/**
+ * This test ensures that RPC calls with unresolved parameters are queued until their parameters are correctly resolved.
+ * This test contains 1 Server and 2 Client workers running in multiple processes.
+ *
+ * The flow is as follows:
+ * - Setup:
+ *  - Launch at least one client in a separate process
+ * - Test:
+ *  - All clients launched outside of the editor process must ensure that the referenced material asset is not in memory at the start of the
+ * test.
+ *  - All clients launched outside of the editor process must successfully resolve the material asset before passing it into a client RPC
+ * function.
+ */
+
+ASpatialTestRPCTimeout::ASpatialTestRPCTimeout()
+	: Super()
+{
+	Author = "Iwan";
+	Description = TEXT(
+		"This test calls an RPC with an asset that was softly referenced and check that it will be asynchronously loaded with a timeout of "
+		"0.");
+}
+
+void ASpatialTestRPCTimeout::CreateCustomContentForMap()
+{
+	GetWorldSettingsForMap()->DefaultGameMode = ASpatialTestRPCTimeoutGameMode::StaticClass();
+}
+
+void ASpatialTestRPCTimeout::PrepareTest()
+{
+	Super::PrepareTest();
+
+	AddStep(
+		TEXT("Check that the material was not initially loaded on non-editor clients"), FWorkerDefinition::AllClients, nullptr, nullptr,
+		[this](float DeltaTime) {
+			if (!GIsEditor)
+			{
+				ACharacter* TestCharacter = Cast<ACharacter>(GetLocalFlowPawn());
+				ASpatialTestRPCTimeoutPlayerController* TestController =
+					Cast<ASpatialTestRPCTimeoutPlayerController>(TestCharacter->GetController());
+
+				if (TestController && TestCharacter)
+				{
+					UMaterial* Material = TestController->SoftMaterialPtr.Get();
+
+					RequireTrue(Material == nullptr, TEXT("The soft-pointed material must not be in memory at the start of the test."));
```

Comment:
hmmm, maybe I would do an AssertTrue, or even better AssertIsNull (I think that exists?) here.
And then make the step take at least 2 seconds or something by doing a RequireTrue(StepTimer > 2seconds).
Right now it will succeed as fast as possible, e.g. on the first instance that the Material is nullptr.

Reply:
This is now addressed and it does work! Though we need to give some latency tolerance; updated the delay to 5 seconds and I require that the material is not loaded for at least the first 4 seconds of the step